### ISSUE TITLE
Running JPype with Cherrypy - Threading

### ISSUE THREAD

**Issue by krowvin** [jpype-project/jpype#1041]
# I'm attempting to run JPype within Cherrypy

I posted this on Stackoverflow as well to try and see if anyone there has gone this route. But I thought who better to ask then the JPype community directly. 

Below is a copy of this [SO Post](https://stackoverflow.com/questions/71711599/using-jpype-with-cherrypy-how-to-handle-threads)

Since i'm in the jpype github, I'll mention, my colleague thinks it has to do with how JPype handles the Java Classes Cache. I could not find any reference of this in the documentation. I can only assume it does not cache the classes for future use? 

## For completeness
I am running this on Solaris/Unix using Python 3.7.10

## Questions:
 1. What is the best practice for this? (Do you recommend global jpype start i.e. outside of all methods, or a cherrypy Tool/Plugin?)
 3. Why would running my java class before calling cherrypy's start make it work, but if I don't call the java class first I get an error on consecutive calls?
 
## What I've Tried:
When I run `jpype.startJVM()` before starting Cherrypy with `cherrypy.engine.start()` I am able to access the jars I imported with Jpype with minimal errors. 

If I run the same jpype script without Cherrypy I get no errors. 

*I get different errors depending on when I run startJVM()*

**If I start the jvm before everything in the global scope - the error looks like this but the class runs**

```
getImageIcon: Error loading Images/Point.gif Error: java.lang.NullPointerException
java.lang.NullPointerException
```

 where getImageIcon is a function of my java class. You would think this file was missing, but if I run the jpype script without cherrypy I do not see this error. 

**If I try to wedge jpype into a plugin/call the class after cherrypy has started I get something like** 
```
plot = PlotClass.createPlot()
java.lang.java.lang.NullPointerException: java.lang.NullPointerException
```

I suspect this has to do with the way cherrypy handles the threads and that Jpype (I think) only attaches to one thread. By running the class calls before cherrypy, jpype properly gets attached to the main thread(?). It's a bit over my head and i'm having trouble finding what I need in the documentation of both. 

## If I output the threads RIGHT before I call the java class with:
```
for thread in threading.enumerate(): 
    print(thread.name)
```

## I get 
```
MainThread
Autoreloader
HTTPServer Thread-2
('CP Server Thread-3',)
('CP Server Thread-4',)
('CP Server Thread-5',)
('CP Server Thread-6',)
('CP Server Thread-7',)
('CP Server Thread-8',)
('CP Server Thread-9',)
('CP Server Thread-10',)
('CP Server Thread-11',)
('CP Server Thread-12',)
```
Not sure how to tell what thread Jpype is running on. 

I tried using Plugins/Tools from Cherrypy and that was a mess. 

## Here's a sample of what i'm working with currently:

*I'm also using SQLAlchemy in a Plugin/Tool in the same script(works great). But have removed that for simplicity.*
```
import cherrypy
import jpype
import jpype.imports
from jpype.types import *

jpype.startJVM(classpath=["my/path/jar/*"])
from myproj.script import PlotClass

class MyProj:
    def __init__(self):
       # If I plot in here when it's initialized it sort of works
       self.plot()
    def plot(self, data):
        for thread in threading.enumerate(): 
            print(thread.name)
        plot = PlotClass.createPlot(data) 
        plot.save("myplot.png")
        if os.path.isfile("myplot.png"):
            # I don't have control over the java classes to make them export a b64 string
            with open("myplot.png", "rb") as image_file:
                encoded_string = base64.b64encode(image_file.read())
            return encoded_string.decode('utf-8')

class Root(object):
    @cherrypy.tools.json_in()
    @cherrypy.tools.json_out()
    @cherrypy.expose
    def plot(self, *args, **kwargs):
        if cherrypy.request.method != 'POST':
            return {"msg": "Must be HTTP POST", "status": "error"}
        data = cherrypy.request.json
        try:
            plot_b64 = mp.plot(data)
            if (plot_b64):
                return {"data": plot_b64, "status": "ok"}
            else:
                return {"data": "Plot not generated, check server logs", "status": "error"}
        except Exception as err:
            print("Unable to plot ", err)
            cherrypy.response.status = 500
            return {"msg": "Failed to plot<br>" + str(err), "status": "error"}
    

if __name__ == '__main__':
    mp = MyProj()
    cherrypy.tree.mount(Root(), '/', config="myproj.conf")
    cherrypy.engine.start()
    cherrypy.engine.block()
    if jpype.isJVMStarted():
        jpype.shutdownJVM()
    
```

# I call the endpoint ip:port/plot and it runs the plotting method for cherrypy

**Comment by krowvin** [comment]
This was more of a question than an issue. Not an issue with JPype to be sure. I'm still working on this but do not want to clog up the existing, real, issues.

**Comment by Thrameos** [comment]
Sound like this could be a problem with paths.  Though you would need to do more diagnostics.   It is important to remember that both virtual machines have different ideas of what the current path is.   Java does bind to only one thread as the primary, but it should latch on to any thread that is used to access Java after starting.  Thus there is not one dominate thread that you need to keep track off (other that when the system shuts down).  

Perhaps you should call some code like
```
from java.nio.file import Paths
print(Paths.get('.').toAbsolutePath())
print(Paths.get('Images/Point.gif').toAbsolutePath())
```

That should help you to resolve if your Java path and your Python paths disagree.

**Comment by krowvin** [comment]
Hello Thrameos, 

I did try running the path code you have above, but I think it's checking at a higher level than the resources I mention are being requested. 
`Paths.get('.').toAbsolutePath()` returns my python scripts current path. 

I imagine these jars are using a relative path within the Java source. I have searched the system I am on and do not see these resources in the filesystem.

The java source I am using is not my own and was provided from a third party. I am currently asking them how they call it and where it resides on the system if at all.

They load fine when I call JPype by itself, but when I try to access the class from within a Cherrypy thread I get the NullPointerException error mentioned. Which makes me think it is on the system somewhere, or that it's generated/drawn (the gifs) only when the java is ran. You mentioned how Java latches, that's good to know. I've moved called Jpype's startJVM into a plugin for Cherrypy. 

Thank you for pointing out the VMs have different ideas of what the current path is. I was not aware of that, I assumed they somehow shared the path it was executed from. This is almost certainly my issue at this point, just a matter of figuring out what "Point.gif" is and where it resides because it's not in my absolutepath/images/point.gif



I'll report back here if I find anything useful. But it seems the issue is on my end.

**Comment by Thrameos** [comment]
The other possibility is that the request may somehow be going through the jpype DynamicClassLoader which may be confused about the resources.   Generally this should only happen when a class library is loaded after the JVM is started, but I wanted you to be aware it is a possibility.   

You can change the Java path using the "user.dir" property in Java as changing paths in Python does not affect the JVM.  

It would also be a good idea to clarify if this is a resource (something stored in the Jar) or something stored in the file system.   You may also want to use the ``ex.stacktrace()`` method to give you a better idea what methods are being called.    

If all else fails you may be able to use the java disassembler to break apart the 3rd party code such that you can try to replicate the exact source of the error, or use the java ASM library which can add new code to an existing library.   Give it works with out cherrypy and fails with, there must be something going on under the hood in the enviroment.   But not knowing what variables the third party code is using, it is not clear how to proceed.   Unless some code is doing something horrible like a "fork()" this should just work.

**Comment by krowvin** [comment]
> DynamicClassLoader which may be confused about the resources 

This might be it! Because I can call the plot method RIGHT after I call jpype and it works. But once cherrypy kicks in if I try to call it using the bus/any other way I get the errors looking for these on-the-fly generated files.  (I confirmed they do not exist in the filesystem, but are rendered as needed. Seems they are snapshots of the current plot state in an icon form.)

I was able to decompile the class being called at the end of the stack trace. This looks the be the offending lines
```
ClassLoader cl = Thread.currentThread().getContextClassLoader();
URL url = cl.getResource(name);
```

I imagine getContextClassLoader() is also using the DynamicClassLoader? I'd also be curious what `Thread.currentThread()` returns. 

I tried setting the user.dir with -Duser.dir and confirmed it was changed with `System.getProperty("user.dir")` and did not see any difference. I suspect because these files are not on the file system. I also tried seeing the path to something that had a `Images` directory in it. 

Going to look into what Thread.currentThread() might return. 

Running ```
jpype.startJVM()
plot = PlotClass.createPlot()
```
Works with no errors within the same scope. It isn't until I hand everything over to cherrypy's bus mechanism that I see theses errors. That's why I was pressed on the thread business so hard. 

I added my Plugin code in an edit in the stack overflow post

**Comment by Thrameos** [comment]
It is not hard to instrument the DynamicClassLoader and recompile org.jpype.jar.  If you place the modified org.jpype.jar in your classpath it will use it instead of the internal copy that is part of the jpype module.  I use this method to diagnose issues in the class loader as well as testing new internals methods.  You can then debug or fix the error and submit a pull request to jpype so that it is addressed for others.

**Comment by krowvin** [comment]
Okay, I will see about modifying the org.jpype.jar for my case. 

I decided to run `print(Thread.currentThread())` in various places. In all of the startup places before CherryPy it shows
```
Thread[main,5,main]
```

When I run it using the pub/sub plugin after CherryPy has started (and I get the errors) I see:
```
1st run: Thread[Thread-3,5,main]
2nd run: Thread[Thread-4,5,main]
3rd run: Thread[Thread-5,5,main]

```

CherryPy reserves a pool of 10 or so threads and uses them on a as needed basis afaik. So it appears it's just using those threads as needed, but the thread with the classes needed for plotting the URI is in the main thread. 

I checked, all the errors I'm getting have classes with the ClassLoader line listed above in it.
